server/controllers/User.js

```diff
@@ -0,0 +1,181 @@
+  * @param {Object} res response object
+  * @returns {Response} response object
+  */
+  getUsers(req, res) {
+    let searchKey = '%%';
+    if (req.query.q) {
+      searchKey = `%${req.query.q}%`;
+    }
+
+    return models.User.findAll({
+      offset: req.query.offset || 0,
+      limit: req.query.limit || 100,
+      attributes: ['id', 'username', 'fullName', 'email', 'roleId', 'about'],
+      where: { username: {
+        $iLike: searchKey
+      } },
+      order: [['id', 'ASC']]
+    })
+    .then(users => res.status(200).send(users))
+    .catch(error => res.status(400).send(error));
+  },
+
+  /**
+  * Create a user
+  * Route: POST: /users
+  * @param {Object} req request object
+  * @param {Object} res response object
+  * @returns {Response} response object
+  */
+  create(req, res) {
+    if (req.body.roleId === '1') {
+      return res.status(401).send({ message: 'Invalid roleId' });
+    }
+    req.body.password =
+      bcrypt.hashSync(req.body.password, bcrypt.genSaltSync(10));
+
+    return models.User.create(req.body)
+      .then((user) => {
+        const token = authenticator.generateToken({
+          id: user.id,
+          username: user.username,
+          roleId: user.roleId
+        });
+        const response = authenticator.secureUserDetails(user);
+        response.message = 'User created';
+        response.token = token;
+        return res.status(201).send(response);
+      })
+      .catch(error => res.status(400).send(error));
+  },
+
+  /**
+  * Get a user
+  * Route: GET: /users/:id
+  * @param {Object} req request object
+  * @param {Object} res response object
+  * @returns {Response} response object
+  */
+  getUser(req, res) {
+    return models.User.findById(req.params.id)
+      .then((user) => {
+        if (!user) return res.status(404).send({ message: 'User not found' });
+
+        res.status(200).send(authenticator.secureUserDetails(user));
+      })
+      .catch(error => res.status(400).send(error));
+  },
+
+  /**
+  * Update a user
+  * Route: PUT: /users/:id
+  * @param {Object} req request object
+  * @param {Object} res response object
+  * @returns {Response} response object
+  */
+  update(req, res) {
+    if (req.body.roleId === '1' && res.locals.decoded.roleId !== 1) {
+      return res.status(403).send({
+        message: 'Only an admin can upgrade a user to an admin role'
+      });
+    }
+    if (req.body.password) {
+      req.body.password =
+      bcrypt.hashSync(req.body.password, bcrypt.genSaltSync(10));
+    }
+
+    return res.locals.user.update(req.body, { fields: Object.keys(req.body) })
+      .then(updatedUser =>
+        res.status(200).send(authenticator.secureUserDetails(updatedUser)))
+      .catch(error => res.status(400).send(error));
+  },
+
+  /**
+  * Delete a user
+  * Route: DELETE: /users/:id
+  * @param {Object} req request object
+  * @param {Object} res response object
+  * @returns {Response} response object
+  */
+  delete(req, res) {
```

Comment:
Can an `Admin` delete himself? If so, who then administers the system after the `adminz is deleted?